tests/http/test_proxyagent.py

```diff
@@ -451,6 +452,36 @@ def test_https_request_via_uppercase_proxy_with_blacklist(self):
         body = self.successResultOf(treq.content(resp))
         self.assertEqual(body, b"result")
 
+    @patch.dict(os.environ, {"http_proxy": "proxy.com:8888"})
```

Comment:
Should we be adding a test for connecting to a proxy via HTTPS proxy or is that already covered by other tests?

Reply:
We aren't *exactly* testing this end-to-end no. I tried going that way with the first iteration but the existing proxy tests are way above my level of twisted internals knowledge already and this test would be a bunch more involved as it would mean nesting 2 levels of fake tls connections inside each other :-/ (It's certainly doable but I can't figure it out, sorry.)

I feel this would be about 100 times easier to test when real connections through a real proxy server. (but doing this in CI brings of course all manner of different headaches.)

One thing to note: This only *actual* change regarding the connection setup in this MR is the addition of those two lines:

```
        tls_options = tls_options_factory.creatorForNetloc(host, port)
        proxy_endpoint = wrapClientTLS(tls_options, proxy_endpoint)
```

Which is the same thing that happens basically in all other places where we do a tls connection from synapse. So in a way this is tested already, just not exactly in case of using a proxy. :-/

Reply:
I think it would be worthwhile to test the agent's behaviour end-to-end with an https proxy, and we have pretty much all the tools we need to do this in the test class already.

Starting with `test_http_request_via_proxy`; the modifications needed are:
 * tell the `ProxyAgent` to accept the self-signed certs with `contextFactory=get_test_https_policy()`
 * extend `_make_connection` to tell it to present a cert for a different host (currently it is hardcoded to present a certificate for `test.com`)
 * where `_make_connection` is called, tell it to wrap the server-side connection with TLS (with `ssl=True`)

I ended up writing this, so have just pushed the change to your branch, in 9326de3d67. Please take a look and check that you think it looks sensible.

Reply:
The lower level plumbing bits of this still seem a bit magic too me, but the changes you outlined look completely sensible. Thanks for adding this.